[#975] Convert blurbs to use Modal components.

LMulvey opened the pull request:
# Description 

Converted the contributor blurs to use the `Modal`  React component.

## More Details

Changed current implementation of contributor blurs to open a `Modal` component containing the profile. I'm very open to visual changes here as I'm not sure this is visually up to snuff yet.

## Corresponding Issue

Fixes #975 

# Screenshots

![contributormodal](https://user-images.githubusercontent.com/16218249/46516900-9c855c00-c829-11e8-9c93-681a9944fb93.gif)

# Test Coverage

🚫 

I could not find any evidence of acceptance or integration testing at a view level. This uses a previously tested React component. Happy to write tests if they're required!

nshki commented:
Oooh thanks for working on this! Let me know if you have any questions regarding the design! You should be able to see the modal design in a frame called "Contributors / Bio" in the Figma doc.

If you haven't used Figma before, you'll be able to pull component properties by clicking things, and you'll be able to export SVGs yourself as well. :)

LMulvey commented:
Ah, that Figma link is fantastic, thank you! 
I'll start on improving this to match the design and push my changes for another review in a bit.

julianguyen commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,9 @@
+<div class="gridManyItem center">
+  <%= react_component('Modal', props: {
+    element: { component: 'Avatar', props: { src: ProfilePicture.normalize_url(blurb["image"]), name: blurb['name'], alt: "Contributor: #{blurb['name']}"} },
```
Ooh! This is a good workaround 🙌 

Since we support i18n, we should have translations for that contributor text, maybe it could be, "Contributor biography for %{name}".

You can go ahead as use Google translate for now and we'll have the appropriate people review the translations later :P

julianguyen commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,9 @@
+    body: blurb["profile"]
+  }) %>
+
+ 
```
Let's remove these extra lines

julianguyen commented on app/views/pages/contribute.html.erb:
```diff
@@ -1,5 +1,6 @@
 <% title t('navigation.contribute') %>
 
+
```
Let's remove this extra line

julianguyen commented on client/app/stories/Modal.stories.jsx:
```diff
@@ -43,4 +44,12 @@ storiesOf('Modal', module)
       openListener={openListener}
     />
+  ))
+  .add('Avatar component as the element', () => (
```
Let's have test coverage for this case (in Avatar.spec.jsx)

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -97,23 +98,53 @@ export class Modal extends React.Component<Props, State> {
-  render() {
+  resolveElement = () => {
     const { element, elementId } = this.props;
+    let Component = null;
```
You don't have to initialize these as null. I think you can just do `let Component;` and `let renderComponent;`.

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -97,23 +98,53 @@ export class Modal extends React.Component<Props, State> {
+    if (typeof element === 'object' && element.component) {
+      const { component, props } = element;
+      Component = this.resolveComponent(component);
+      renderComponent = <Component {...props} />;
```
You could alternatively use `React.createElement()` to create your Avatar component and pass in the props.

julianguyen commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,9 @@
+  <%= react_component('Modal', props: {
+    element: { component: 'Avatar', props: { src: ProfilePicture.normalize_url(blurb["image"]), name: blurb['name'], alt: "Contributor: #{blurb['name']}"} },
+    title: nil,
+    body: blurb["profile"]
```
We still want to have:
```
<i class="fa fa-location-arrow fa-inline"></i><%= " #{blurb["location"]}" %>
        <br /><%= content_tag(:i, '', class: "#{get_icon_class(blurb["social"])} fa-inline") %><%= link_to(blurb["link_name"], blurb["link"], target: 'blank') %>
```

You can pass a partial in this case and it won't cause problems cause there's no nested React component haha.

LMulvey replied:
Ah, I will cover this all off. Still a WIP :D

LMulvey commented:
I'm close here but not yet done! Going to re-work the modal design a bit to match the Figma comps. Should be finished up tomorrow. Feel free to keep reviewing as I push changes, though @julianguyen 😀

julianguyen commented:
Now that we're using React for this, we can delete the following:
- app/assets/javascripts/contributors_readmore.js
- spec/javascripts/ContributorsReadMoreSpec.js

🎉 

One step closer to no more files in the Rails JS asset pipeline 🙏

julianguyen commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,21 @@
+      } 
+    },
+    title: nil,
+    body: <<~HEREDOC
```
TIL

julianguyen commented on client/app/components/Modal/__tests__/Modal.spec.jsx:
```diff
@@ -196,5 +196,23 @@ describe('Modal', () => {
       });
     });
+
+    describe('uses an Avatar component for an element and HTML values for the body', () => {
```
Thanks for adding tests :D

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -97,23 +98,53 @@ export class Modal extends React.Component<Props, State> {
+      renderComponent = React.createElement(
+        this.resolveComponent(component),
+        { ...props },
+        null,
```
Why is null being passed?

LMulvey replied:
I'm actually brand new to the `createElement` API and figured I _had_ to pass a value to the `children` argument, even if I didn't explicitly want children.

julianguyen replied:
Aha, thanks for explaining. I think you can actually exclude that third argument entirely (it appears to be an optional value). I did that for client/app/mocks/InputMocks.jsx and it worked :P

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -97,23 +98,53 @@ export class Modal extends React.Component<Props, State> {
+    if (element) {
+      return (
+        <div
+          id={elementId || null}
```
We can actually get rid of the null condition since elementId can be undefined or a string.

LMulvey commented:
Alright! Pushed my latest work up. Here's some screenshots:

![hackerman screencap 2018-10-12 at 11 57 08 pm](https://user-images.githubusercontent.com/16218249/46902004-ca4a5080-ce7a-11e8-84fa-2a969dfd1eaa.png)
![hackerman screencap 2018-10-12 at 11 57 17 pm](https://user-images.githubusercontent.com/16218249/46902005-ca4a5080-ce7a-11e8-877c-7b3b6d55857e.png)

Few notes:
1) Added new SVG close button. I chose to integrate it right into the Modal component. Not sure if we want to look at having a wrapper component for custom SVG icons that aren't FA... but short term, I felt this was the best solution. I don't know where else would be using the X SVG icon.

2) Cleaned up the files mentioned by @julianguyen  - this included some stale SCSS that I found in `~app/assets/stylesheets/application/contributors.scss` - replaced with my own SCSS rules for the newer designs.

As always, let me know what changes you want/need but I feel like we're real close here! 👍

julianguyen commented on app/assets/stylesheets/application/contributors.scss:
```diff
@@ -1,9 +1,34 @@
-.hideContent {
-  display: none;
-}
+.contributorModal-body {
```
We haven't had a consensus on using BEM, though I think it's awesome. For now, can you change these class names to camel case? Happy to discuss more outside of this PR :)

julianguyen commented on client/app/components/Modal/Modal.scss:
```diff
@@ -52,7 +52,9 @@
       }
       &Close {
         cursor: pointer;
-        margin-left: auto;
```
Let's not use position: absolute here. It's not necessary.

LMulvey replied:
![without](https://user-images.githubusercontent.com/16218249/48751925-26f01680-ec44-11e8-8540-f346fa53616c.png)
Without position absolute, the close button takes up 15px at the top of the modal _as if there were a title_ there. With position absolute, I can have the close button at the top-right without pushing the rest of the content down since I'm _not_ using the title prop of the modal component here.

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -53,12 +54,29 @@ export class Modal extends React.Component<Props, State> {
+      fill="#6D0839"
+      xmlns="http://www.w3.org/2000/svg"
+    >
+      <path
```
We should save this as an SVG file under `app/assets/images/`.

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -53,12 +54,29 @@ export class Modal extends React.Component<Props, State> {
 
+  displayCloseSvg = () => (
+    <svg
+      width="36"
```
We should use CSS to set the size of this icon and make it responsive.

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -1,10 +1,9 @@
 // @flow
 import React from 'react';
 import renderHTML from 'react-render-html';
-import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
```
We'll also want to replace the close icon in the Header component as well to be consistent. You can use CSS to change the fill colour of it.

LMulvey replied:
Reverted the SVG back to FAIcons as per discussion below.

julianguyen commented on config/locales/de.yml:
```diff
@@ -358,6 +358,7 @@ de:
     blog:
       description: 'Wenn du für uns schreiben möchtest, stelle dich gerne per E-Mail bei uns vor: %{email}! Auch anonyme Einsendungen werden berücksichtigt.'
     contribute:
+      contributor_avatar_alt: "Mitwirkende Biographie für %{name}"
```
Just to be consistent with our linting, let's use single quotes instead of double quotes.

nshki commented on client/app/components/Modal/Modal.scss:
```diff
@@ -22,15 +22,16 @@
-    width: 50vw;
-    height: 60vh;
+    width: 60vw;
+    height: 65vh;
```
Also a great start. I think it would be great to have the modal container expand to fit all the content, and then let the user scroll to see more if it overflows past the visible viewport.

LMulvey replied:
100% agree! A bunch of this stuff was pre-existing and I'm super gun-shy when making modifications to existing code on a project I don't own. Maybe I should trust my own gut a little more 👍

LMulvey replied:
To expand on these comments and the ones above here, I'm thinking that we should have a max-height and max-width set + overflow: auto. Give the modal a little room to breathe responsively but also not get disproportionately long height-wise before hitting the final breakpoint and going fullscreen.

LMulvey replied:
I messed around with the idea you presented–allowing the modal height to expand based on the content height and then scrolling the window over the overlay. It presents a host of new issues to correct unfortunately as flexbox positions the element in the center of the window right now based on the center point of the element, and not based on the top-left corner. 

So, essentially, when the content-height expands beyond the height of the screen, the top of the modal disappears and is unreachable. We can correct and reposition it, but we'd have to do that based on content-height and using JS, I think. Maybe @julianguyen has another solution! I've left it intact right now, giving it a max-height to match the Figma comp (as well as the max-width previously discussed above).

LMulvey replied:
I messed around with the idea you presented–allowing the modal height to expand based on the content height and then scrolling the window over the overlay. It presents a host of new issues to correct unfortunately as flexbox positions the element in the center of the window right now based on the center point of the element, and not based on the top-left corner. 

So, essentially, when the content-height expands beyond the height of the screen, the top of the modal disappears and is unreachable. We can correct and reposition it, but we'd have to do that based on content-height and using JS, I think. Maybe @julianguyen has another solution! I've left it intact right now, giving it a max-height to match the Figma comp (as well as the max-width previously discussed above).

nshki commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,34 @@
+            #{blurb["name"]}
+          </h4>
+          <div class="contributorModal-body-socialLinks">
+            <i class="fa fa-location-arrow fa-inline"></i>#{blurb["location"]}
```
Let's reduce the space to the right of this element so it matches the first icon in this section.

nshki commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,34 @@
+<div class="gridManyItem center">
```
Would love to tune up the responsive behavior of this component. 😄

When the viewport is <= 640px, I like that you made the modal fill the entire screen, but it would also be super cool if the avatar sat on its own line (centered, I think), the name, location, and link were centered, and the bio kept left-aligned.

julianguyen commented:
@nshki What do you think of just using the Font Awesome close icon instead of the custom one? I think it's to leverage that library since we're already using it.

nshki commented:
@julianguyen Hmm not opposed to it! However, I think ideally we'd eventually switch off of icon fonts completely and use custom SVGs so we have free reign over behavior and animations.

julianguyen commented:
@nshki You can still apply CSS (including for animations) to the Font Awesome icon fonts. Other than animations, what behaviour would we want?

nshki commented:
@julianguyen Hmm true. I guess I'm only thinking about animations atm. e.g. Moving lines to transform a hamburger icon into a close icon.

julianguyen commented:
For now, let's just stick with the original Font Awesome close icon. We can revisit all of our SVG icons on another issue.

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

LMulvey commented:
Pushed up a bunch more changes this evening, take a look when you both get chances. I feel like we're close now! I addressed all the comments above a took a better look at what happens when we hit the small breakpoint to make it look nicer. 

Also, replaced the SVG icon with the FA icon as per above comments. I left the icon positioned absolutely for now as per my [reasoning above](https://github.com/ifmeorg/ifme/pull/1098#discussion_r226531469).

As always, let me know your thoughts and I'll make changes as needed! Excited to get this out 😄

nshki commented on client/app/components/Modal/Modal.scss:
```diff
@@ -22,15 +22,18 @@
     @include setFontSize($size-20);
     @include fadeIn(0.5s);
-    padding: $size-20;
+    padding: $size-30 $size-50;
```
Let's change this to 35px in all directions so the spacing around the scrollable area looks more evenly distributed.

nshki commented on app/assets/stylesheets/application/contributors.scss:
```diff
@@ -1,9 +1,49 @@
+  }
+
+	.contributorModalBodyCopy {
+    margin-left: $size-18;
```
I think this can be removed so we can define spacing around the modal content in the parent box element.

LMulvey replied:
I've left this in for now. Without it, the content is crammed up against the avatar at larger breakpoints. Not sure what you meant by defining spacing around the modal content in the parent box element.
![nospacing](https://user-images.githubusercontent.com/16218249/48752521-ed6cda80-ec46-11e8-8a96-1714e80f240a.png)

nshki commented on app/assets/stylesheets/application/contributors.scss:
```diff
@@ -1,9 +1,49 @@
+    
+    .contributorModalHeader {
+      display: flex;
+      align-items: center;
```
At smaller breakpoints, would love to stack the avatar, name, and social links on top of one another, so perhaps a `flex-direction: column;` would be appropriate here! Reverting to `flex-direction: row;` for larger breakpoints is totally fine, to keep it looking like how it did before.

nshki commented on app/assets/stylesheets/application/contributors.scss:
```diff
@@ -1,9 +1,49 @@
+        margin: $size-24 0 $size-10 0;
+        @media screen and (max-width: $small) {
+          margin-top: 0;
+        }
```
It would be great if we can add in `text-align: center;` for smaller breakpoints.

nshki commented on app/views/pages/_contributor_blurb.html.erb:
```diff
@@ -0,0 +1,43 @@
+                <i class="fa fa-location-arrow fa-inline"></i>#{blurb["location"]}
+                #{content_tag(:i, '', class: "#{get_icon_class(blurb["social"])} fa-inline")}
+                #{link_to(blurb["link_name"], blurb["link"], target: 'blank')}
+              </div>
```
On smaller breakpoints, it would be great to separate out both location and social link onto their own lines, so perhaps wrapping each in a `<div>` would be good. You can then style them to be happy with each other on the same line as the screen gets larger.

julianguyen commented:
@LMulvey Thanks for the update and taking the time to work on this :) No need to apologize - it's all part of the development process.

nshki commented:
Ditto what Julia said 100%. No need to apologize! Thanks a ton for working on this. 😄

julianguyen commented:
Hey @LMulvey, any updates on this? :)

LMulvey commented:
Sorry for the wait again! I've officially dropped one of my part-time commitments to work on more projects like this. As always, let me know what further changes you want and my turn around should be quicker.

julianguyen commented:
No worries! Thanks for the update :)

julianguyen commented on app/assets/stylesheets/application/contributors.scss:
```diff
@@ -1,9 +1,90 @@
+      border-radius: 50%;
+      overflow: hidden;
+      object-fit: cover;
+      background: linear-gradient(104.26deg, #6d0839 0%, #D0E799 175.81%);
```
I think you should be able to use `$mulberry` and `$key-lime` for the respective hex values.

julianguyen commented on client/app/components/Modal/Modal.scss:
```diff
@@ -22,15 +22,18 @@
     @include setFontSize($size-20);
     @include fadeIn(0.5s);
-    padding: $size-20;
+    padding: $size-35;
```
Let's just round up and use `$size-36` :P

julianguyen commented on client/app/components/Modal/index.jsx:
```diff
@@ -97,23 +100,51 @@ export class Modal extends React.Component<Props, State> {
     const { element, elementId } = this.props;
+    let renderComponent;
+
+    if (typeof element === 'object' && element.component) {
```
I think using `element && element.component` would suffice.
